pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -1039,6 +1044,126 @@ private void internalUnloadTransactionCoordinatorAsync(AsyncResponse asyncRespon
+
+        CompletableFuture<Void> ret;
+        if (topicName.isGlobal()) {
+            ret = validateGlobalNamespaceOwnershipAsync(namespaceName);
+        } else {
+            ret = CompletableFuture.completedFuture(null);
+        }
+        ret.thenAccept(__ -> {
+            long ledgerId = deleteLedgerPayload.getLedgerId();
+            log.info("[{}][{}] received delete ledger: {}", clientAppId(), topicName, ledgerId);
+            validateTopicOwnershipAsync(topicName, authoritative)
+                    .thenCompose(ignore ->
+                            //Is need to check delete_ledger operation?
+                            validateTopicOperationAsync(topicName, TopicOperation.DELETE_LEDGER))
+                    .thenCompose(ignore -> getTopicReferenceAsync(topicName))
+                    .thenCompose(topic -> {
+                        CompletableFuture<Void> future = new CompletableFuture<>();
+                        if (topic == null) {
+                            asyncResponse.resume(new RestException(Status.NOT_FOUND,
+                                    getTopicNotFoundErrorMessage(topicName.toString())));
+                            future.complete(null);
+                            return future;
+                        }
+                        ManagedLedger managedLedger = ((PersistentTopic) topic).getManagedLedger();
+                        DeleteLedgerPayload.OffloadContext context = deleteLedgerPayload.getOffloadContext();
+                        LedgerType ledgerType = LedgerType.valueOf(deleteLedgerPayload.getLedgerType());
+                        if (LedgerType.LEDGER == ledgerType) {
+                            managedLedger.asyncDeleteLedger(topicName.getPersistenceNamingEncoding(), ledgerId,
+                                    ledgerType, null).whenComplete((res, ex) -> {
+                                        if (ex != null) {
+                                            if (ex instanceof PendingDeleteLedgerInvalidException) {
+                                                if (log.isDebugEnabled()) {
+                                                    log.debug("[{}][{}] Received invalid pending delete ledger {},"
+                                                                    + " invalid reason: {}", clientAppId(), topicName,
+                                                            ledgerId, ex.getMessage());
+                                                }
+                                                future.complete(null);
+                                                return;
+                                            }
+                                            future.completeExceptionally(ex);
+                                            return;
+                                        }
+                                        future.complete(null);
+                                    });
+                        } else if (LedgerType.OFFLOAD_LEDGER == ledgerType) {
```

Comment:
Can we merge these two cases into one? The main differences are the `offloadContext` parameter. In `managedledger.asyncDeleteLedger` method, we also checked the `ledgerType`

Reply:
Yes, it can be.